### PR TITLE
Warm start for cbify

### PR DESCRIPTION
**Pull request by zcc1307**
This patch is for a new mode in vw: contextual bandit learning with warm start (CB-WS). It is mainly based on modifying cbify.cc - in its predict_or_learn_adf, the learning is now broken to two phases: 1. warm start phase and 2. interaction phase.

Remarks:
1. As of now, the CB-WS mode only works if the base exploration algorithm is epsilon-greedy. For more complex exploration algorithms (e.g. cover/bagging), a difficulty is that, we need to initialize all the base learners in cover/bagging using warm start examples - this seems to require use to change the code of predict_or_learn_cover/predict_or_learn_bag in cb_explore_adf.cc.

2. We additionally scale the cb examples' importance weight by a 1/num_actions in the mtr mode of cb_adf.cc - this has the effect of ensuring the warm start examples have the same weight as the CB examples.  

3. Fixed an offset issue in predict_or_learn_greedy (should use example's offset - otherwise will have the wrong behavior when multiple cb_explore learners are initialized in cbify) and multiline_predict_or_learn (store / restore the examples' offsets properly)

4. Added some simple test cases.

5. There are lots of debugging cout's - perhaps I should delete them?

### CODE REVIEW

**Comment by JohnLangford** [comment]
Chicheng, can you update to the current master and resolve conflicts?

**Comment by JohnLangford** [comment]
This pull request **introduces 7 alerts** when merging 5fa45cb0a39905e79fd53e61f9b158c7bae97f05 into c4bb9bcd59e8c7834a10c5d9d64c212792e647ff - [view on LGTM.com](https://lgtm.com/projects/g/JohnLangford/vowpal_wabbit/rev/pr-3aa3d6ebf204ee03438b74129906c8348e8493c2)

**new alerts:**

* 3 for Unused import
* 2 for Unused local variable
* 1 for Unreachable code
* 1 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by JohnLangford** [comment]
This pull request **introduces 7 alerts** when merging 2020c50730e93913632fb4a0ca215b628b70cdf4 into a300157538f7568b5c980625f6df979931bd7a15 - [view on LGTM.com](https://lgtm.com/projects/g/JohnLangford/vowpal_wabbit/rev/pr-5c16cb1bb96c7b93cb557c85178637cdb5a01900)

**new alerts:**

* 3 for Unused import
* 2 for Unused local variable
* 1 for Unreachable code
* 1 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by JohnLangford** [comment]
This pull request **introduces 7 alerts** when merging f95d1541b1a17576fd4e97998ff2a11f65b747cf into a300157538f7568b5c980625f6df979931bd7a15 - [view on LGTM.com](https://lgtm.com/projects/g/JohnLangford/vowpal_wabbit/rev/pr-e26eb142ca2b096c125f6e4a214a094c74a46ef2)

**new alerts:**

* 3 for Unused import
* 2 for Unused local variable
* 1 for Unreachable code
* 1 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by zcc1307** [comment]
Additional issues:
1. We don't have a label corruption scheme for cost-sensitive labels yet. Right now, if use_cs is true, we simply do not perform label corruption.
2. For diagnostic purposes, warm_cb outputs average variance estimate, theoretical average variance, and last lambda chosen when it finishes. Shall I remove them? (It does not affect the test results though, since they go to stdout.)

**Comment by JohnLangford** [comment]
Chicheng, can you resolve the conflicts on this?  I'd like to merge.   

There have been two substantial changes since this was written: the options parser was abstracted and we switched to using cmake.   If you grab @jackgerrits for a few minutes he should be able to guide updates relatively quickly.

File: vowpalwabbit/cb_explore_adf.cc [hunk 1]
```diff
@@ -260,6 +260,7 @@ void predict_or_learn_first(cb_explore_adf& data, multi_learner& base, multi_ex&
 template <bool is_learn>
 void predict_or_learn_greedy(cb_explore_adf& data, multi_learner& base, multi_ex& examples)
 {
+	data.offset = examples[0]->ft_offset;
```
Comment (JohnLangford) [inline vowpalwabbit/cb_explore_adf.cc]: Why do we set something which is never used?

Comment (zcc1307) [reply vowpalwabbit/cb_explore_adf.cc]: I believe we need this additional line - we had an email discussion on this on July 7, 2018:

If we don't have this line, in warm_cb.cc, base.learn(data.adf_data.ecs, 0) / base.learn(data.adf_data.ecs, 1)'s updates seem to be on the same set of weights.

**Comment by zcc1307** [comment]
Some other thoughts:
1. Shall we remove the implementations for warm-start ground truth (including options on corruptions on interaction data, per-dataset weighting scheme, and warm-start validation for lambda selection)? 

- First, it is not evaluated in our paper. 
- Second, the current implementation of per-dataset weighting scheme is problematic. Ideally, we would like for each round t, (total weight of warm start data) / (total weight of interaction data) = (1-lambda) / lambda. However, right now, (total weight of warm start data) / (total weight of interaction data) = (1 - lambda) / (lambda * t / T), where T is the total number of examples in the interaction stage.


2. We also have a Sim-Bandit implementation in warm_cb.cc - this can be triggered by setting --warm_start_type to 2 - shall we keep it?

3. We also tried using \epsilon_t (time-varying)-greedy exploration (although not reflected in this pull request) - shall I add it?

**Comment by zcc1307** [comment]
Some more questions:

1. I think test case 173 fails because we are scaling the importance weight of each example in the MTR reduction by 1/num_actions - shall I change the test result reference instead?

On the other hand, I am confused about the implementation of cost regression in csoaa - in csoaa, is each example (x,c) converted to a loss \ell(f, (x,c)) = \sum_{a=1}^K (f(x,a) - c(a))^2, or \ell(f, (x,c)) = 1/K * \sum_{a=1}^K (f(x,a) - c(a))^2? 

If it is the former, then I think we shouldn't scale the loss in the MTR reduction by 1/K, as before scaling, \EE[\hat{\ell}(f, (x,c))] = \sum_{a=1}^K (f(x,a) - c(a))^2. (Basically we would like the optimization objective to be exactly the same as the one in [our paper](https://arxiv.org/pdf/1901.00301.pdf), Appendix A.)

2. Are we happy with showing the VW doubling progress report with number of examples starting from #(warm start examples)? (see e.g. the first VW output in [this page](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Warm-starting-contextual-bandits)) In my implementation, I set the examples' importance weight to zero if they are not in the interaction stage, so that we are only counting the average loss in the interaction stage.

3. We explicitly use --warm_start_update and --interaction_update as two input options to indicate if VW turn on updates in the respective two stages. Is it too long for users?

**Comment by jackgerrits** [comment]
@zcc1307 I'm going to help you get this in. Grab me at some point and let me know how I can help here.

File: vowpalwabbit/cb_adf.cc [hunk 2]
```diff
@@ -111,10 +111,8 @@ void learn_MTR(cb_adf& mydata, multi_learner& base, multi_ex& examples)
   gen_cs_example_mtr(mydata.gen_cs, examples, mydata.cs_labels);
   uint32_t nf = (uint32_t)examples[mydata.gen_cs.mtr_example]->num_features;
   float old_weight = examples[mydata.gen_cs.mtr_example]->weight;
-  examples[mydata.gen_cs.mtr_example]->weight *= 1.f / examples[mydata.gen_cs.mtr_example]->l.cb.costs[0].probability *
-      ((float)mydata.gen_cs.event_sum / (float)mydata.gen_cs.action_sum);
-  GEN_CS::call_cs_ldf<true>(
-      base, mydata.gen_cs.mtr_ec_seq, mydata.cb_labels, mydata.cs_labels, mydata.prepped_cs_labels, mydata.offset);
+  examples[mydata.gen_cs.mtr_example]->weight *= 1.f / examples[mydata.gen_cs.mtr_example]->l.cb.costs[0].probability * ((float)mydata.gen_cs.event_sum / (float)mydata.gen_cs.action_sum) * (1.f / (float)examples.size());
+  GEN_CS::call_cs_ldf<true>(base, mydata.gen_cs.mtr_ec_seq, mydata.cb_labels, mydata.cs_labels, mydata.prepped_cs_labels, mydata.offset);
```
Comment (zcc1307) [inline vowpalwabbit/cb_adf.cc]: Sorry, I am a little confused about this line - what is ((float)mydata.gen_cs.event_sum / (float)mydata.gen_cs.action_sum)?  It is 1/K?

**Comment by zcc1307** [comment]
@jackgerrits Thanks, Jack! Yes, I think I have made changes to pass the checks. 

@JohnLangford As we further divide the importance weights in the mtr reduction by 1/K, this caused some changes in regcb's test results. I am not sure if I should change the Lambda set setting as we discussed before to accomodate this 1/K change in mtr, because we also hope that the algorithm works also for other reductions (ips/dr)?

Instructions on using the --warm_cb option can be found in https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Warm-starting-contextual-bandits .

**Comment by JohnLangford** [comment]
@zcc1307 I'm kind of confused.   What is the question exactly?

We have some conflicts to merge.

File: vowpalwabbit/cb_adf.cc (hunk 2 above)
Comment (JohnLangford) [reply vowpalwabbit/cb_adf.cc]: It's more like 1 / average K.  These are defined via this: https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/vowpalwabbit/gen_cs_example.cc#L171

**Comment by JohnLangford** [comment]
Is this ready to merge?

**Comment by zcc1307** [comment]
Yes, it is ready to merge! I have also updated the documentations:
https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Warm-starting-contextual-bandits

**Comment by jackgerrits** [comment]
Thanks for the docs! I’ll add it to the sidebar today

**Comment by JohnLangford** [comment]
Merged, thanks :-)

**Pull request merged**